homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,262 @@
+    )
+    platform.async_register_entity_service(
+        SERVICE_SET_ONGRID_BATTERY_DOD,
+        {vol.Required(ATTR_ONGRID_BATTERY_DOD): vol.Coerce(int)},
+        "async_set_ongrid_battery_dod",
+    )
+    platform.async_register_entity_service(
+        SERVICE_SET_GRID_EXPORT_LIMIT,
+        {vol.Required(ATTR_GRID_EXPORT_LIMIT): vol.Coerce(int)},
+        "async_set_grid_export_limit",
+    )
+
+    return True
+
+
+class InverterEntity(CoordinatorEntity, SensorEntity):
+    """Entity representing the inverter instance itself."""
+
+    _MAIN_ENTITY_SENSOR = "ppv"
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        device_info: DeviceInfo,
+        inverter: Inverter,
+    ) -> None:
+        """Initialize the main inverter entity."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_inverter"
+        self._attr_unique_id = f"{DOMAIN}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_icon = "mdi:solar-power"
+        self._attr_name = "PV Inverter"
+        self._attr_native_unit_of_measurement = POWER_WATT
+        self._attr_extra_state_attributes = {"serial_number": inverter.serial_number}
+        self._inverter: Inverter = inverter
+
+    async def async_set_work_mode(self, work_mode: int) -> None:
+        """Set the inverter work mode."""
+        await self._inverter.set_work_mode(work_mode)
+
+    async def async_set_ongrid_battery_dod(self, ongrid_battery_dod: int) -> None:
+        """Set the on-grid battery dod."""
+        await self._inverter.set_ongrid_battery_dod(ongrid_battery_dod)
+
+    async def async_set_grid_export_limit(self, grid_export_limit: int) -> None:
+        """Set the grid export limit."""
+        await self._inverter.set_grid_export_limit(grid_export_limit)
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self.coordinator.data is not None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        if self.coordinator.data is not None:
+            new_value = self.coordinator.data.get(self._MAIN_ENTITY_SENSOR)
+            # If no new value was provided, keep the previous
+            if new_value is not None:
+                self._attr_native_value = new_value
+
+        return self._attr_native_value
+
+    @property
+    def state_attributes(self):
+        """Return the inverter state attributes."""
+        return self.coordinator.data
+
+
+class InverterSensor(CoordinatorEntity, SensorEntity):
+    """Entity representing individual inverter sensor."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        device_info: DeviceInfo,
+        inverter: Inverter,
+        sensor: Sensor,
+    ) -> None:
+        """Initialize an inverter sensor."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_{sensor.id_}"
```

Comment:
It is not common to have `entity_id` defined, most components use the default.
If you do choose to define it, then the format looks weird to me (starting with a dot???).

Reply:
Well, I copied from somewhere at very beginning of integration development. It is used in other integrations as well and without the dot at the beginning it will crash in `split_entity_id` method call in entity_platform.py.
Second, as mentioned above in `_attr_unique_id `, any change here will break existing data of many users, so unless really required, I'd prefer to maintain backwards compatibility

Reply:
`split_entity_id` probably crashes because the platform is missing. Entity ids are usually in the form "{platform}.{domain}.{name}". I very much doubt that "master" reviewers will accept entities without a platform (even if it's a breaking change from the original custom component).

Reply:
As I mentioned above, I very likely copied this from some existing integration and it produces entity names like `sensor.goodwe_e_total` which seems to be perfectly legal name.
I'm open to any suggestion how to implement entity naming strategy differently (prefferrably keeping existing names).
I've found this way and don't know any better.

Reply:
Setting a name is enough. It will be the template for the entity_id.

Reply:
I see. Setting the name is sufficient but the generated entity ID will be a bit too verbose.
At this moment it e.g. generates entity name "fgrid3" with friendly name "On-grid L3 Frequency".
Not setting entity_id explicitly would generate entity name on_grid_l3_frequency.
The idea at the beginning was the entity friendly name could be in theory localizable, while the id is technical name of the sensor.
Is not explicitly setting entity_id strongly requested or is it just suggestion and we can keep current behavior ?
(Which would be welcome by community which already collected months of data with these names ...)

Reply:
If there are already entity entries with the same unique id in the entity registry the entities will keep the entity id set in the registry. For new users we can use the new entity id template from the name. They can change it as wanted in the GUI.

Reply:
Explicit entity_id setting was removed.